lib/Doctrine/Enumeration/Enumerated.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Enumeration;
+
+use Doctrine\Enumeration\Exception\NotSupportedException;
+use function array_search;
+use function constant;
+use function in_array;
+
+trait Enumerated
+{
+    final private function __construct()
+    {
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function validateValue($value) : void
+    {
+        if (! in_array($value, self::availableValues(), true)) {
+            throw new NotSupportedException('Invalid value.');
+        }
+    }
+
+    final public static function validateName(string $name) : void
+    {
+        if (! isset(self::availableValues()[$name])) {
+            throw new NotSupportedException('Invalid name.');
+        }
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function get($value) : self
+    {
+        self::validateValue($value);
```

Comment:
Looks like this line does the same as `array_search()` below. Can you instead of calling the same lookup twice just throw an exception if the `$name` is not found?

Reply:
Should it be `$name` and `validateName()` instead of `$value` and `validateValue()`? We're getting elements by name, not by value.

Reply:
Looks like we are getting them by value: `return self::get(self::SERIALIZABLE);`. Might be an issue if the values are not unique.

Reply:
Values are always unique

Reply:
You hope :P

Reply:
We could also rely on the constant names, rather than their values...

Reply:
> We're getting elements by name, not by value.

It's by value. Doing that by name would kill the ability to use constants or compare the instance against constant easily.